pkg/adsc/delta.go

```diff
@@ -0,0 +1,739 @@
+//     - Tree Update: Modifies its 'tree' to reflect changes in resources, such as adding new resources,
+//     updating relationships between parents and children, and removing or unlinking resources.
+//     - Resource Caching: If 'EnableDeletedResourceCache' is enabled, newly added resources are cached.
+//     This cache plays a critical role when resources are subsequently removed from the stream, allowing
+//     the client to trigger delete events with the cached resource. In the absence of caching, delete
+//     events carry only the resource's name.
+//
+//  3. State Synchronization: Post-processing the delta response, the client updates its internal state. This involves:
+//     - Acknowledgements and Errors: Communicating acknowledgements or errors back to the server based on the
+//     processing outcome. In cases of error or rejection, a Nack can be sent using HandlerContext.Reject.
+//     - Dependency Updates: Triggering requests for dependent resources. These dependencies are established via
+//     HandlerContext.RegisterDependency.
+//
+// An example of a handler registration is as follows:
+//
+//	clusterHandler := Register(func(ctx HandlerContext, res *cluster.Cluster, event Event) {
+//	  if event == EventDelete {
+//	    return
+//	  }
+//	  ctx.RegisterDependency(v3.SecretType, ExtractClusterSecretResources(t, res)...)
+//	  ctx.RegisterDependency(v3.EndpointType, ExtractEdsClusterNames([]*cluster.Cluster{res})...)
+//	})
+//
+// It means that when a cluster is added or updated, the client will trigger requests for the
+// secrets and endpoints that the cluster depends on.
+//
+// An example of register handlers:
+//
+//	handlers := []Option{
+//	  clusterHandler,
+//	  Watch[*cluster.Cluster]("*"),
+//	  listenerHandler,
+//	  Watch[*listener.Listener]("*"),
+//	  endpointsHandler,
+//	  routesHandler,
+//	  secretsHandler,
+//	}
+//
+// client := NewDelta("localhost:8080", handlers...)
+//
+// It means that the client will watch all clusters and listeners, and trigger resource events for
+// clusters, listeners, endpoints, routes and secrets that the clusters and listeners depend on.
+type Client struct {
+	cfg      *DeltaADSConfig
+	handlers map[string]HandlerFunc
+	// tree is a map where each key is a `resourceKey` (comprising the resource name and typeURL)
+	// and each added resource is a `resourceNode`. This tree structure represents the dynamic state
+	// and relationships of resources.
+	tree map[resourceKey]resourceNode
+
+	xdsClient discovery.AggregatedDiscoveryService_DeltaAggregatedResourcesClient
+	conn      *grpc.ClientConn
+
+	// initialWatches is the list of resources we are watching on startup
+	initialWatches []resourceKey
+
+	// sendNodeMeta is set to true if the connection is new - and we need to send node meta
+	sendNodeMeta atomic.Bool
+
+	mutex sync.RWMutex
+	// Last received message, by type
+	received        map[string]*discovery.DeltaDiscoveryResponse
```

Comment:
What's this for?

IIRC, it is used for async response. The it is not quite right, because we cannot gurantee the orders.

Reply:
> I would also rename it lastReceived

I think naming this "lastReceived" makes sense? As you said, this doesn't guarantee the order, so users will need to implement custom checks on what was received. Like:

https://github.com/istio/istio/blob/06ad655977d4619343f530280d0afeece9d95945/pkg/adsc/delta_test.go#L437-L454


However, if we don't include this logic, it would require users to implement more complex logic to determine when the response for the watched resources has terminated.